fix solution for Meetup title

eclectic-mind opened the pull request.

eclectic-mind committed:
fix solution for Meetup title

jsru commented:
Проверьте, пожалуйста, ваше решение, не все тесты прошли (PR не будет принят до тех пор, пока **все** добавленные задачи не будут решены).

ShGKme commented on 01-basics/03-MeetupTitle/script.js:
```diff
@@ -2,37 +2,40 @@ import Vue from './vendor/vue.esm.browser.js';
+
   methods: {
     getMeetupData: function () {
+      if (this.number === 0) return;
```
Вместо значения 0 лучше использовать `null`.

eclectic-mind committed:
add solution for Meetup page draft

не удалось доделать: фоновую картинку, иконки, тип пункта программы вместо названия

jsru commented:
Проверьте, пожалуйста, ваше решение, не все тесты прошли (PR не будет принят до тех пор, пока **все** добавленные задачи не будут решены).

ShGKme commented on 01-basics/04-MeetupPageDraft/index.html:
```diff
@@ -43,15 +52,15 @@ <h5 class="meetup-agenda__title">Название доклада</h5>
             <li>
               <img class="icon info-list__icon" alt="icon" src="/assets/icons/icon-cal-lg.svg" />
-              <time datetime="2020-02-31">31 фев. 2020</time>
+              <time v-bind:datetime="datetime">{{ date }}</time>
```
Принято всегда использовать короткую запись `:datetime="datetime"`

ShGKme commented on 01-basics/04-MeetupPageDraft/script.js:
```diff
@@ -44,4 +44,133 @@ const getAgendaItemIcons = () => ({
+    place: function () {
+      if (this.meetup === null) return;
+      return this.meetup.place;
+    },
```
Вместо всех проверок в вычисляемых свойствах, лучше один раз добавить проверку `v-if="meetup"` и не выводить целиком страницу митапа, пока нет данных митапа.

Тогда эти вычисляемые свойства не понадобятся.

ShGKme commented on 01-basics/04-MeetupPageDraft/script.js:
```diff
@@ -44,4 +44,133 @@ const getAgendaItemIcons = () => ({
+    },
+    noAgenda: function () {
+      return (this.meetup === null || this.meetup.agenda === null) ? true : false;
+    },
```
Программа отсутствует, если `meetup.agenda.length === 0`, это проверку можно оставить в шаблоне.

ShGKme commented on 01-basics/03-MeetupTitle/index.html:
```diff
@@ -8,7 +8,7 @@
   <div id="app">
     <form>
-      <label v-for="n in 5"><input type="radio" name="meetup-number" v-bind:value="{n}" v-model="number"> {{ n }}</label>
+      <label v-for="n in 5"><input type="radio" name="meetup-number" v-bind:value="n" v-model="number"> {{ n }}</label>
```
Принято всегда использовать короткую форму `:value="n"`

eclectic-mind committed:
finished: solution for Meetup page draft

eclectic-mind committed:
Merge branch 'master' of https://github.com/js-tasks-ru/vue-20210304_maria-korolkova

eclectic-mind committed:
Merge branch 'master' of https://github.com/js-tasks-ru/vue-20210304_maria-korolkova

eclectic-mind committed:
add solution for Meetup description

jsru commented:
Проверьте, пожалуйста, ваше решение, не все тесты прошли (PR не будет принят до тех пор, пока **все** добавленные задачи не будут решены).

eclectic-mind committed:
update Couner Button Component

jsru commented:
Проверьте, пожалуйста, ваше решение, не все тесты прошли (PR не будет принят до тех пор, пока **все** добавленные задачи не будут решены).

ShGKme commented on 01-basics/04-MeetupPageDraft/index.html:
```diff
@@ -11,28 +11,37 @@
+
+          <p v-if="noAgenda" class="meetup-agenda__empty">Программа пока пуста, но когда-нибудь в ней обязательно что-нибудь появится!</p>
+
+          <div v-else="noAgenda" class="meetup-agenda">
```
У `v-else` нет значения. Как и с `if-else` в js, это просто `else`, который идёт после `v-if`.

ShGKme commented on 01-basics/04-MeetupPageDraft/index.html:
```diff
@@ -11,28 +11,37 @@
+            <div v-for="item in meetup.agenda" class="meetup-agenda__item">
               <div class="meetup-agenda__item-col">
-                <img class="icon" alt="icon" src="/assets/icons/icon-cal-sm.svg" />
+                <img class="icon" alt="icon" v-bind:src="icon(item)" />
```
Лучше избавиться от вызова методов в шаблоне. Здесь может быть простое выражение или вычисляемое свойство.

ShGKme commented on 01-basics/04-MeetupPageDraft/index.html:
```diff
@@ -11,28 +11,37 @@
-                <h5 class="meetup-agenda__title">Название доклада</h5>
-                <p><span>Докладчик</span><span class="meetup-agenda__dot"></span><span class="meetup-agenda__lang">RU</span></p>
-                <p>Описание доклада</p>
+                <h5 v-if="hasTitle(item)" class="meetup-agenda__title">{{ item.title }}</h5>
+                <h5 v-else="hasTitle(item)" class="meetup-agenda__title">{{ showDefaultTitle(item) }}</h5>
+                <p v-if="hasSpeaker(item)">
```
Аналогично:
1. У v-else не может быть значения
2. Лучше избавиться от вызова методов в шаблоне. Шаблон - декларация представления.

ShGKme commented on 01-basics/04-MeetupPageDraft/script.js:
```diff
@@ -44,4 +44,100 @@ const getAgendaItemIcons = () => ({
+        })
+        .then((data) => {
+          this.meetup = data;
+          console.log(data);
```
В будущем лучше убирать отладочный вывод, чтобы он не попадал в результаты тестирования

eclectic-mind committed:
add solution for Meetup description. cover, info, view

jsru commented:
Проверьте, пожалуйста, ваше решение, не все тесты прошли (PR не будет принят до тех пор, пока **все** добавленные задачи не будут решены).

ShGKme commented on 02-components/03-MeetupCover/MeetupCover.js:
```diff
@@ -1,8 +1,28 @@
+  computed: {
+    meetupTitle() {
+      return (this.title === null || !this.itle) ? 'Demo meetup' : this.title;
+    },
```
Какого-то значения по умолчанию `Demo meetup` по условию нет.
Если хочется, чтобы у какого-то параметра было значение по умолчанию, то лучше описать его сразу в свойстве `default` при описании входного параметра.

B в этом коде опечатка `this.itle`

ShGKme commented on 02-components/05-MeetupAgenda/MeetupAgendaItem.js:
```diff
@@ -1,21 +1,59 @@
+    }
+  },
+
+  methods: {
+    showDefaultTitle() {
+      const type = this.agendaItem.type;
+      const obj = getAgendaItemDefaultTitles();
+      const title = obj[type];
+      return title;
+    },
+
+    showIcon() {
+        const type = this.agendaItem.type; 
+        const obj = getAgendaItemIcons();
+        const src = obj[type];
+        return `/assets/icons/icon-${src}.svg`;
+    },
+  },
```
Эти методы нужны только внутри своих вычисляемых свойств. Проще этот код сразу описывать внутри вычисляемого свойства, не создавая отдельные методы

ShGKme commented on 02-components/05-MeetupAgenda/MeetupAgendaItem.js:
```diff
@@ -1,21 +1,59 @@
-        <p>
-          <span>Докладчик</span>
+        <h5 v-if="agendaItem.title" class="meetup-agenda__title">{{ agendaItem.title }}</h5>
+        <h5 v-else class="meetup-agenda__title">{{ defaultTitle }}</h5>
```
Сам HTML элемент заголовка сейчас дублируется, отличается только содержимое.
В этом случае можно или использовать внутри тернарный оператор, или сделать это в вычисляемом свойстве.

`<h5 class="meetup-agenda__title">{{ agendaItem.title ? agendaItem.title : defaultTitle }}</h5>`

Или ещё короче, но менее очевидно:
`<h5 class="meetup-agenda__title">{{ agendaItem.title && defaultTitle }}</h5>`

ShGKme commented on 02-components/06-MeetupView/MeetupView.js:
```diff
@@ -1,26 +1,51 @@
+  computed: {
+    link() {
+      return
+    },
```
Не хватает ссылки...

ShGKme commented on 02-components/06-MeetupView/MeetupView.js:
```diff
@@ -1,26 +1,51 @@
   template: `
     <div>
-      <!-- meetup cover -->
+      <meetup-cover :link="meetup.link" :title="meetup.title" />
```
В митапе нет свойства `link`

eclectic-mind committed:
fix Meetup title

eclectic-mind committed:
fix Meetup page draft

jsru commented:
Проверьте, пожалуйста, ваше решение, не все тесты прошли (PR не будет принят до тех пор, пока **все** добавленные задачи не будут решены).

eclectic-mind committed:
counter button component

jsru commented:
Проверьте, пожалуйста, ваше решение, не все тесты прошли (PR не будет принят до тех пор, пока **все** добавленные задачи не будут решены).

eclectic-mind committed:
add solution for Meetup page

jsru commented:
Проверьте, пожалуйста, ваше решение, не все тесты прошли (PR не будет принят до тех пор, пока **все** добавленные задачи не будут решены).

eclectic-mind committed:
fix: meetup view, agenda, cover

jsru commented:
Проверьте, пожалуйста, ваше решение, не все тесты прошли (PR не будет принят до тех пор, пока **все** добавленные задачи не будут решены).

eclectic-mind committed:
Meetup page

jsru commented:
Проверьте, пожалуйста, ваше решение, не все тесты прошли (PR не будет принят до тех пор, пока **все** добавленные задачи не будут решены).

ShGKme commented on 02-components/06-MeetupView/MeetupView.js:
```diff
@@ -1,26 +1,55 @@
+  },
+
+  computed: {
+    link() {
+      return `https://course-vue.javascript.ru/api/images/${this.meetup.imageId}`;
+    },
```
`this.meetup.imageId` - может быть null, если изображения нет. В этом случае получится некорректная ссылка. Здесь требуется проверка. Кстати, для получения ссылки на изображение по ID есть функция `getImageUrlByImageId` в `data.js`.

ShGKme commented on 02-components/06-MeetupView/MeetupView.js:
```diff
@@ -1,26 +1,55 @@
+    }
+  },
+
+  created() {
+    console.log(this.meetup);
+  },
```
Отладочный вывод потребуется убрать

ShGKme reviewed (commented):
1. В 01-basics/04-MeetupPageDraft остались проблемы с использованием методов в шаблоне по первым комментариям. 
2. В 02-components/01-CounterButtonComponent вы не сообщаете родительскому компоненту о том, что значение увеличилось на 1. Требуется породить событие с новым значением. 

По остальным задачам оставил комментарии по коду.

eclectic-mind committed:
counter button

jsru commented:
Проверьте, пожалуйста, ваше решение, не все тесты прошли (PR не будет принят до тех пор, пока **все** добавленные задачи не будут решены).

eclectic-mind committed:
fix Meetup page

jsru commented:
Проверьте, пожалуйста, ваше решение, не все тесты прошли (PR не будет принят до тех пор, пока **все** добавленные задачи не будут решены).

eclectic-mind committed:
clean code (components)

jsru commented:
Проверьте, пожалуйста, ваше решение, не все тесты прошли (PR не будет принят до тех пор, пока **все** добавленные задачи не будут решены).

ShGKme commented on 02-components/07-MeetupPageComponent/MeetupPage.js:
```diff
@@ -1,8 +1,49 @@
+
+  template: 
+  `<div v-if="meetup">
+    <meetup-view 
+    :date="meetup.date" 
+    :agenda="meetup.agenda" 
+    :type="meetup.type" 
+    :title="meetup.title" 
+    :place="meetup.place" 
+    :description="meetup.description" 
+    :organizer="meetup.organizer" 
+    :meetup="meetup" />
+  </div>`,
```
У `meetup-view` только один параметр - `meetup`

eclectic-mind committed:
fix solution for Counter button component

jsru commented:
Проверьте, пожалуйста, ваше решение, не все тесты прошли (PR не будет принят до тех пор, пока **все** добавленные задачи не будут решены).

eclectic-mind committed:
fix solution for Meetup page

jsru commented:
Проверьте, пожалуйста, ваше решение, не все тесты прошли (PR не будет принят до тех пор, пока **все** добавленные задачи не будут решены).

eclectic-mind committed:
fix solution for Meetup page draft

jsru commented:
Проверьте, пожалуйста, ваше решение, не все тесты прошли (PR не будет принят до тех пор, пока **все** добавленные задачи не будут решены).

ShGKme commented on 02-components/01-CounterButtonComponent/CounterButton.js:
```diff
@@ -28,6 +46,6 @@ export default {
 
-  template: '<button type="button" @click="increment">{{ count }}</button>',
-};
+  template: '<button type="button" @click="increment" v-model="count">{{ count }}</button>',
```
У кнопки не может быть модели... Модель - это **двустороннее** связывание, которое актуально для элементов/компонентов, в который вводится или выбирается какое-то значение. Поля ввода, радиокнопки и т.д.

`v-model="count"` - это сокращённая запись для двух директив: `:attr="count" @event="count = $event"`, где attr - это атрибут (параметр) элемента, который определяет значение, а `event` - событие **на элементе" которое происходит при его изменении в элементе (например, в поле ввода это событие input).

У кнопки хоть и есть атрибут value, но нет события, с котором оно изменяется.

ShGKme commented on 01-basics/04-MeetupPageDraft/index.html:
```diff
@@ -10,29 +10,38 @@
-                <h5 class="meetup-agenda__title">Название доклада</h5>
-                <p><span>Докладчик</span><span class="meetup-agenda__dot"></span><span class="meetup-agenda__lang">RU</span></p>
-                <p>Описание доклада</p>
+                <h5 v-if="hasTitle(item)" class="meetup-agenda__title">{{ item.title }}</h5>
+                <h5 v-else class="meetup-agenda__title">{{ showDefaultTitle(item) }}</h5>
```
Здесь остались методы. При этом в каждом из методов выполняется всего одна операция. Лучше оставить эту операцию явно прямо в шаблоне, не вынося в отдельный метод.

eclectic-mind committed:
fix solutions

jsru commented:
Проверьте, пожалуйста, ваше решение, не все тесты прошли (PR не будет принят до тех пор, пока **все** добавленные задачи не будут решены).

ShGKme commented on 01-basics/04-MeetupPageDraft/index.html:
```diff
@@ -34,8 +34,8 @@ <h3>Программа</h3>
-                <h5 v-if="hasTitle(item)" class="meetup-agenda__title">{{ item.title }}</h5>
-                <h5 v-else class="meetup-agenda__title">{{ showDefaultTitle(item) }}</h5>
+                <h5 v-if="item.title !== null" class="meetup-agenda__title">
+                  {{ item.title !== null ? item.title : getAgendaItemDefaultTitles()[item.type] }}</h5>
```
Чтобы `getAgendaItemDefaultTitles` каждый раз не запускать (объект один и тот же), можно было один раз его результат сохранить в данных приложения

ShGKme commented on 01-basics/04-MeetupPageDraft/index.html:
```diff
@@ -34,8 +34,8 @@ <h3>Программа</h3>
-                <h5 v-else class="meetup-agenda__title">{{ showDefaultTitle(item) }}</h5>
+                <h5 v-if="item.title !== null" class="meetup-agenda__title">
+                  {{ item.title !== null ? item.title : getAgendaItemDefaultTitles()[item.type] }}</h5>
                 <p v-if="item.type === 'talk' ? true : false">
```
Использование тернарного оператора в варианте `item.type === 'talk' ? true : false` бесполезно.
Дословно, это получается "если Х - истинно, тогда истинно, иначе (если ложно), тогда ложно".

`v-if="item.type === 'talk' ? true : false"` - тоже самое, что `v-if="item.type === 'talk'"`. Аналогично в обычных `if`.

ShGKme reviewed (commented):
Принято 👍

eclectic-mind committed:
Merge branch 'master' of https://github.com/js-tasks-ru/vue-20210304_maria-korolkova

eclectic-mind committed:
add solution for Counter button.vue

jsru commented:
Проверьте, пожалуйста, ваше решение, не все тесты прошли (PR не будет принят до тех пор, пока **все** добавленные задачи не будут решены).

ShGKme commented:
`CounterButton.vue` принята

eclectic-mind committed:
add solution for FormGroup

jsru commented:
Проверьте, пожалуйста, ваше решение, не все тесты прошли (PR не будет принят до тех пор, пока **все** добавленные задачи не будут решены).

eclectic-mind committed:
Merge branch 'master' of https://github.com/js-tasks-ru/vue-20210304_maria-korolkova

eclectic-mind committed:
Merge branch 'master' of https://github.com/js-tasks-ru/vue-20210304_maria-korolkova

ShGKme commented on 03-sfc/02-FormGroup/components/FormGroup.vue:
```diff
@@ -1,13 +1,26 @@
-  <div class="form-group">
-    <label class="form-label">Label</label>
-    <!-- <input /> -->
+  <div class="form-group" :class="['form-group_inline', { 'inline': true }]">
```
При такой записи `form-group_inline` будет добавлен всегда. Также класс `inline` тоже будет добавлен всегда, так как его значение - true.

Здесь же требуется добавлять только класс `form-group_inline` только, когда `inline` истино.

https://vuejs.org/v2/guide/class-and-style.html#Object-Syntax

eclectic-mind committed:
fix solution for FormGroup

eclectic-mind committed:
add solution for DropdownButton

jsru commented:
Проверьте, пожалуйста, ваше решение, не все тесты прошли (PR не будет принят до тех пор, пока **все** добавленные задачи не будут решены).

eclectic-mind committed:
fix solution for DropdownButton: src in styles

jsru commented:
Проверьте, пожалуйста, ваше решение, не все тесты прошли (PR не будет принят до тех пор, пока **все** добавленные задачи не будут решены).

eclectic-mind committed:
fix solution for DropdownButton

jsru commented:
Проверьте, пожалуйста, ваше решение, не все тесты прошли (PR не будет принят до тех пор, пока **все** добавленные задачи не будут решены).

ShGKme commented on 03-sfc/02-FormGroup/components/FormGroup.vue:
```diff
@@ -1,13 +1,33 @@
+    }
+  },
+
+  methods: {
+    isInline() {
+      return this.inline;
+    }
+  }
```
Этот метод не нужен, он ничего не делает. Можно обращаться напрямую к `inline`.

ShGKme commented on 03-sfc/02-FormGroup/components/FormGroup.vue:
```diff
@@ -1,13 +1,33 @@
-    <label class="form-label">Label</label>
-    <!-- <input /> -->
+  <!-- <div class="form-group" :class="['form-group_inline', { 'inline': true }]"> -->
+    <div class="form-group" :class="{ isInline: form-group_inline }">
```
С точки зрения JS тут написано, что есть объект со свойством isInline, значение которого: "переменная form минус переменная group_inline". Так как таких переменных нет, это `undefined вычесть undefined`.

https://vuejs.org/v2/guide/class-and-style.html#Object-Syntax

ShGKme commented on 03-sfc/04-DropdownButton-1/components/DropdownButton.vue:
```diff
@@ -26,7 +36,213 @@ export default {
+    
+    options: {
+      type: Array,
+      default: function() {
+        return [
+          value: {
+            type: String,
+            required: true
+          },
+          text: {
+            type: String,
+            required: true
+            },
+          icon: {
+            type: String,
+            required: false
+            }
+          ],
+        }
```
Значение по умолчанию - это значение, которое будет у входного параметра, если его значение не передано. 

Здесь также ошибка синтаксическая. В JS нельзя описать массив со свойствами как [ value: {} ].
Это может быть либо массив объектов, либо объект с объектами.

В любом случае, здесь бы это означало, что при отсутствии `options` в нём должно быть какое-то значение со свойствами внутри type и разными объектами типов внутри. Это явно не может быть списком вариантов.

ShGKme commented on 03-sfc/04-DropdownButton-1/components/DropdownButton.vue:
```diff
@@ -26,7 +36,213 @@ export default {
+  },
+
+  methods: {
+    hasIcon() {
+      return this.options.length > 2;
+    },
```
Иконка есть, если у хотя бы одного варианта из массива `options` есть поле `icon`.
Здесь поможет метод массива `some`: https://developer.mozilla.org/ru/docs/Web/JavaScript/Reference/Global_Objects/Array/some

ShGKme commented on 03-sfc/04-DropdownButton-1/components/DropdownButton.vue:
```diff
@@ -26,7 +36,213 @@ export default {
+    hasIcon() {
+      return this.options.length > 2;
+    },
+    changeValue(item) {
+      if (item.value !==  this.value) {
+        item.value = this.value;
+      }
+    }
```
При выборе значения, компонент должен передавать выбор родительскому компоненту через порождение события. `item` менять нельзя, так как это косвенно входной параметр.

ShGKme commented on 03-sfc/04-DropdownButton-1/components/DropdownButton.vue:
```diff
@@ -26,7 +36,213 @@ export default {
+  top: 15px;
+  right: 16px;
+  transform: none;
+  background: url('@/assets/icons/icon-chevron-down.svg') no-repeat;
```
Здесь должен быть путь `~@/assets...`

ShGKme commented on 03-sfc/04-DropdownButton-1/components/DropdownButton.vue:
```diff
@@ -1,19 +1,29 @@
+        {{ item.text }}
       </button>
-      <button class="dropdown__item dropdown__item_icon" type="button">
+      <!-- <button class="dropdown__item dropdown__item_icon" type="button" :class={ !!hasIcon: 'dropdown__item_icon' }>
```
Объект `{ !!hasIcon: 'dropdown__item_icon' }` некорректный по синтаксису JS. Имя объекта не может быть с операторами `!`.

eclectic-mind committed:
fix solution for FormGroup

jsru commented:
Проверьте, пожалуйста, ваше решение, не все тесты прошли (PR не будет принят до тех пор, пока **все** добавленные задачи не будут решены).

eclectic-mind committed:
fix solution for DropdownButton-1

jsru commented:
Проверьте, пожалуйста, ваше решение, не все тесты прошли (PR не будет принят до тех пор, пока **все** добавленные задачи не будут решены).

ShGKme commented on 03-sfc/02-FormGroup/components/FormGroup.vue:
```diff
@@ -1,13 +1,28 @@
-    <label class="form-label">Label</label>
-    <!-- <input /> -->
+  <!-- <div class="form-group" :class="['form-group_inline', { 'inline': true }]"> -->
+    <div class="form-group" :class="{ 'form-group_inline': !!inline }">
```
Двойное применение оператора отрицания `!!` иногда применяют как короткий аналог приведения к Boolean (другой вариант `Boolean(value)`). Здесь этого не требуется, так как `inline` уже имеет значение типа `boolean`, и приводить `true` к `true` не требуется

ShGKme commented on 03-sfc/02-FormGroup/components/FormGroup.vue:
```diff
@@ -1,13 +1,28 @@
+  props: {
+    label: {
+      type: String,
+      default: null
```
Для строки чаще значением по умолчанию удобнее иметь пустую строку, если вообще требуется `''`. `null` обычно применяют к не примитивным значениям - объектам

ShGKme commented on 03-sfc/04-DropdownButton-1/components/DropdownButton.vue:
```diff
@@ -26,7 +33,219 @@ export default {
+  props: {
+    title: {
+      type: String,
+      default: null,
+      required: true,
```
`default` и `required` обычно не нужны вместе. `required: true` означает, что обязательно должно быть передано значение параметра. А `default` нужен для случаев, когда значение не передано.

ShGKme reviewed (commented):
Все решённые задачи приняты
